poc_iot_injector/src/server.rs

```diff
@@ -139,85 +142,89 @@ async fn submit_txns(
-                                        txn_details.txn
-                                    )
-                                    .await;
-                                }
-                            } else {
-                                tracing::info!("only storing txn: {:?}", txn_details.hash_b64_url);
-                                metrics::increment_counter!("poc_injector_receipt_count");
-                                let _ = file_sink_write!(
-                                    "signed_poc_receipt_txn",
-                                    receipt_sender,
-                                    txn_details.txn
-                                )
-                                .await;
-                            }
-                        }
-                        Err(err) => tracing::warn!("unable to construct txn_details, {err:?}"),
-                    },
+    let mut stream =
+        store.source_unordered(LOADER_WORKERS, stream::iter(file_list).map(Ok).boxed());
+
+    let mut set = JoinSet::new();
+
+    while let Some(msg) = stream.next().await {
+        match msg {
+            Err(err) => tracing::warn!("skipping entry in stream: {err:?}"),
+            Ok(buf) => {
+                let shared_key = keypair.clone();
+                let mut shared_txn_service = txn_service.clone();
+                let receipt_sender_clone = receipt_sender.clone();
+
+                tracing::info!("Spawning submission tasks...");
+                set.spawn(async move {
+                    let _ = process_submission(
+                        buf,
+                        shared_key,
+                        do_submission,
+                        &receipt_sender_clone,
+                        &mut shared_txn_service,
+                    )
+                    .await;
+                });
+
+                if set.len() > MAX_CONCURRENT_SUBMISSIONS {
+                    tracing::info!("Processing {MAX_CONCURRENT_SUBMISSIONS} submissions");
+                    set.join_next().await;
                 }
             }
-        })
-        .await;
+        }
+    }
+
+    // Make sure the tasks are finished to completion even when we run out of stream items
+    while !set.is_empty() {
+        set.join_next().await;
+    }
+
     Ok(())
 }
 
-#[derive(thiserror::Error, Debug)]
-#[error("txn submission failed: {0}")]
-pub struct HandleTxnSubmissionError(pub String);
+async fn process_submission(
+    buf: BytesMut,
+    shared_key: Arc<Keypair>,
+    do_submission: bool,
+    receipt_sender: &file_sink::MessageSender,
+    txn_service: &mut Option<TransactionService>,
+) -> anyhow::Result<()> {
+    match handle_report_msg(buf, shared_key) {
+        Ok(txn_details) => {
+            tracing::debug!("txn_details: {:?}", txn_details);
+            if do_submission {
+                tracing::info!("submitting txn: {:?}", txn_details.hash_b64_url);
+                match handle_txn_submission(txn_details.clone(), txn_service).await {
```

Comment:
do we not want to continue to write the receipts to the output bucket (for testing or auditing at least) when we're submitting them to the chain?

Reply:
Two reasons:
1. I wasn't really sure where it was bottle-necking so I removed the file_sink_write.
2. Its kinda pointless anyway since the receipts would already be on chain and whatever audit, testing, observability is required can be done from chain. Note that we can only be best-effort here as far as submission is concerned.

That said, if you think its still valuable to put the receipts txns on chain when submission is happening, I can add that back in.

cc: @Vagabond

Reply:
i wonder if it would be helpful for correlating submitted txns vs. minted txns but we could always infer the expected submissions from the other upstream data in the pipeline that fed into those receipts.